### PR TITLE
Allow deletion of disappeared groups

### PR DESCRIPTION
**Pull request by bastjan**
This PR allows deletion of disappeared groups.

It does so by checking for the `group-sync-operator.redhat-cop.io/sync-provider` label and deleteing groups which sync timestamp was not updated.

A `static` provider was added for easier testing.

### CODE REVIEW

File: controllers/groupsync_controller.go [hunk 1]
```diff
@@ -174,12 +177,40 @@ func (r *GroupSyncReconciler) Reconcile(context context.Context, req ctrl.Reques
...
+					constants.SyncProvider: providerLabel,
+				}),
+			})
+			if err != nil {
+				log.Error(err, "Failed to list groups for deletion")
+				return r.wrapMetricsErrorWithMetrics(prometheusLabels, context, instance, err)
+			}
+			for _, group := range groupsByProvider.Items {
+				if group.Annotations[constants.SyncTimestamp] == syncTimestampAnnotation {
+					continue
```
Comment (bastjan) [inline controllers/groupsync_controller.go:194]: On second thought maybe we should create a set of seen groups in the steps before? Would make the code a tad more complicated but could be safer.

Comment (glrf) [reply controllers/groupsync_controller.go:194]: If you mean by that to the groups with the correct `providerLabel` before and compare them the `groups` returned by the `groupSyncer` and remove all that are not present, then yes that sound more resilient and I would prefer that.

I'm not quite sure why you are working with the timestamp at all. You have the list of groups that have been returned by the provider and can compare it to the existing groups.

File: controllers/groupsync_controller.go [hunk 2]
```diff
@@ -100,7 +101,9 @@ func (r *GroupSyncReconciler) Reconcile(context context.Context, req ctrl.Reques
 		prometheusLabels := prometheus.Labels{METRICS_CR_NAMESPACE_LABEL: instance.GetNamespace(), METRICS_CR_NAME_LABEL: instance.GetName(), METRICS_PROVIDER_LABEL: groupSyncer.GetProviderName()}
 
 		// Provider Label
-		providerLabel := fmt.Sprintf("%s_%s", instance.Name, groupSyncer.GetProviderName())
+		providerLabel := fmt.Sprintf("%s_%s_%s", instance.Namespace, instance.Name, groupSyncer.GetProviderName())
```
Comment (glrf) [inline controllers/groupsync_controller.go:104]: I get why you're doing that, but that's strictly speaking a breaking change. I have no idea if anyone depends on that label. If we want to upstream this sometime, I would be cautious and use a new label.

File: controllers/groupsync_controller.go [hunk 3]
```diff
@@ -100,7 +101,9 @@ func (r *GroupSyncReconciler) Reconcile(context context.Context, req ctrl.Reques
...
+
+		syncTimestampAnnotation := clock.Now().Format(time.RFC3339Nano)
```
Comment (glrf) [inline controllers/groupsync_controller.go:106]: Technically their `ISO8601` is not RFC3339. But their implementation is utterly broken so fixing that is fine I guess :sweat_smile:

File: controllers/groupsync_controller_test.go [hunk 4]
```diff
@@ -0,0 +1,127 @@
...
+
+	AfterEach(func() {
+		ctx := context.Background()
+
+		k8sClient.DeleteAllOf(ctx, &redhatcopv1alpha1.GroupSync{}, client.InNamespace(namespace))
+		k8sClient.DeleteAllOf(ctx, &userv1.User{})
+		k8sClient.DeleteAllOf(ctx, &userv1.Group{})
+	})
+
+	It("It should sync groups from the provider", func() {
```
Comment (glrf) [inline controllers/groupsync_controller_test.go:64]: For Ginkgo tests you don't need to add the `It` or `By` as in the output that should be added already.  But that's a very minor nitpick :)
```suggestion
	It("should sync groups from the provider", func() {
```

**Review by glrf — CHANGES_REQUESTED** [review]
Looks pretty good. I really like the tests :)

However I don't understand why you are using timestamps. Maybe I'm missing something. My approach would have been

* Get all OpenShift groups with the provider label
* Get the list of groups from the provider
* Make diff of these two lists and remove all that are in the former but not in the latter

That would be more resilient and IMO not more complicated.

File: README.md [hunk 5]
```diff
@@ -452,6 +452,30 @@ The secret can be created by executing the following command:
...
+  name: static-sync
+spec:
+  providers:
+    - name: static
+      static:
+        groups:
+          - name: developers
+            users: [xerxes.xoe, mermes.moe]
+```
+
```
Comment (ccremer) [inline README.md:478]: It would have been better to make the static provider into its own PR, but no hard feelings here.

**Review by ccremer — APPROVED** [review]
I agree with @glrf 's suggestions

File: README.md (hunk 5 above)
Comment (bastjan) [reply README.md:478]: I used the static provider for the unit tests too. that's why it's an all-in-one PR 😅

File: controllers/groupsync_controller.go (hunk 2 above)
Comment (bastjan) [reply controllers/groupsync_controller.go:104]: You're right. I'll add an additional label instead.

File: controllers/groupsync_controller.go [hunk 6]
```diff
@@ -101,6 +102,9 @@ func (r *GroupSyncReconciler) Reconcile(context context.Context, req ctrl.Reques
 
 		// Provider Label
 		providerLabel := fmt.Sprintf("%s_%s", instance.Name, groupSyncer.GetProviderName())
+		providerNamespaceLabel := instance.Namespace
+
+		syncTimestampAnnotation := clock.Now().Format(time.RFC3339)
```
Comment (glrf) [inline controllers/groupsync_controller.go:107]: As you now don't need the timestamp to be the same for all groups, I would personally move that to down to where it's used and not add another variable.

File: controllers/groupsync_controller.go [hunk 7]
```diff
@@ -174,12 +179,38 @@ func (r *GroupSyncReconciler) Reconcile(context context.Context, req ctrl.Reques
...
+				log.Error(err, "Failed to list groups for deletion")
+				return r.wrapMetricsErrorWithMetrics(prometheusLabels, context, instance, err)
+			}
+
+			deletedGroups := findDeletedGroups(groupsByProvider.Items, groups)
+			deletedGroupsCount = len(deletedGroups)
+			for _, group := range deletedGroups {
+				err := r.GetClient().Delete(context, &group)
+				if err != nil {
+					log.Error(err, "Failed to Create or Update OpenShift Group")
```
Comment (glrf) [inline controllers/groupsync_controller.go:201]: Missed that last time. You're not creating or updating anything :sweat_smile:

**Review by glrf — APPROVED** [review]
LGTM :rocket: 

Only some very minor nitpicks.

File: controllers/groupsync_controller.go (hunk 7 above)
Comment (bastjan) [reply controllers/groupsync_controller.go:201]: 🖨🍝

**Pull request merged**